homeassistant/components/mqtt/__init__.py

```diff
@@ -57,6 +59,22 @@
 MAX_RECONNECT_WAIT = 300  # seconds
 
 
+# Service call validation schema
+def mqtt_topic(value):
+    """Validate that we can publish using this MQTT topic."""
+    if isinstance(value, str) and all(c not in value for c in '#+\0'):
+        return vol.Length(min=1, max=65535)(value)
+    raise vol.Invalid('Invalid MQTT topic name')
+
+MQTT_PUBLISH_SCHEMA = vol.Schema({
+    vol.Required(ATTR_TOPIC): mqtt_topic,
+    vol.Exclusive(ATTR_PAYLOAD, 'payload'): object,
+    vol.Exclusive(ATTR_PAYLOAD_TEMPLATE, 'payload'): cv.string,
+    vol.Required(ATTR_QOS, default=DEFAULT_QOS): vol.In([0, 1, 2]),
+    vol.Required(ATTR_RETAIN, default=DEFAULT_RETAIN): vol.Coerce(bool),
```

Comment:
Good point. I'll fix the tests and coerce QOS

On Sat, Apr 2, 2016 at 9:06 PM, Jan Harkes <notifications@github.com> wrote:

> In homeassistant/components/mqtt/__init__.py
> <https://github.com/balloob/home-assistant/pull/1683#discussion_r58304169>
> :
>
> > @@ -57,6 +59,22 @@
> >  MAX_RECONNECT_WAIT = 300  # seconds
> >
> >
> > +# Service call validation schema
> > +def mqtt_topic(value):
> > +    """Validate that we can publish using this MQTT topic."""
> > +    if isinstance(value, str) and all(c not in value for c in '#+\0'):
> > +        return vol.Length(min=1, max=65535)(value)
> > +    raise vol.Invalid('Invalid MQTT topic name')
> > +
> > +MQTT_PUBLISH_SCHEMA = vol.Schema({
> > +    vol.Required(ATTR_TOPIC): mqtt_topic,
> > +    vol.Exclusive(ATTR_PAYLOAD, 'payload'): object,
> > +    vol.Exclusive(ATTR_PAYLOAD_TEMPLATE, 'payload'): cv.string,
> > +    vol.Required(ATTR_QOS, default=DEFAULT_QOS): vol.In([0, 1, 2]),
> > +    vol.Required(ATTR_RETAIN, default=DEFAULT_RETAIN): vol.Coerce(bool),
>
> I think I had tried cv.Boolean, but had tests fail. But those may have
> been failures caused by something else. Also if you assume inputs might be
> strings ATTR_QOS probably should get coerced to int as well.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/1683/files/35dd1aced7eec7888e839b779dddbe25b0297893#r58304169>
>



-- 
PaulusSchoutsen.nl
It's nice to be important but it's more important to be nice.